backend/common/src/http_utils.rs

```diff
@@ -0,0 +1,172 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use futures::io::{BufReader, BufWriter};
+use hyper::server::conn::AddrStream;
+use hyper::{Body, Request, Response, Server};
+use std::future::Future;
+use std::net::SocketAddr;
+use tokio_util::compat::{Compat, TokioAsyncReadCompatExt};
+
+/// A convenience function to start up a Hyper server and handle requests.
+pub async fn start_server<H, F>(addr: SocketAddr, handler: H) -> Result<(), anyhow::Error>
+where
+    H: Clone + Send + Sync + 'static + FnMut(SocketAddr, Request<Body>) -> F,
+    F: Send + 'static + Future<Output = Result<Response<Body>, anyhow::Error>>,
+{
+    let service = hyper::service::make_service_fn(move |addr: &AddrStream| {
+        let mut handler = handler.clone();
+        let addr = addr.remote_addr();
+        async move { Ok::<_, hyper::Error>(hyper::service::service_fn(move |r| handler(addr, r))) }
+    });
+    let server = Server::bind(&addr).serve(service);
+
+    log::info!("listening on http://{}", server.local_addr());
+    server.await?;
+
+    Ok(())
+}
+
+type WsStream = BufReader<BufWriter<Compat<hyper::upgrade::Upgraded>>>;
+pub type WsSender = soketto::connection::Sender<WsStream>;
+pub type WsReceiver = soketto::connection::Receiver<WsStream>;
+
+/// A convenience function to upgrade a Hyper request into a Soketto Websocket.
+pub fn upgrade_to_websocket<H, F>(req: Request<Body>, on_upgrade: H) -> hyper::Response<Body>
+where
+    H: 'static + Send + FnOnce(WsSender, WsReceiver) -> F,
+    F: Send + Future<Output = ()>,
+{
+    if !is_upgrade_request(&req) {
+        return basic_response(400, "Expecting WebSocket upgrade headers");
+    }
+
+    let key = match req.headers().get("Sec-WebSocket-Key") {
+        Some(key) => key,
+        None => {
+            return basic_response(
+                400,
+                "Upgrade to websocket connection failed; Sec-WebSocket-Key header not provided",
+            )
+        }
+    };
+
+    if req
+        .headers()
+        .get("Sec-WebSocket-Version")
+        .map(|v| v.as_bytes())
+        != Some(b"13")
+    {
+        return basic_response(
+            400,
+            "Sec-WebSocket-Version header should have a value of 13",
+        );
+    }
+
+    // Just a little ceremony we need to go to to return the correct response key:
+    let mut accept_key_buf = [0; 32];
+    let accept_key = generate_websocket_accept_key(key.as_bytes(), &mut accept_key_buf);
+
+    // Tell the client that we accept the upgrade-to-WS request:
+    let response = Response::builder()
+        .status(hyper::StatusCode::SWITCHING_PROTOCOLS)
+        .header(hyper::header::CONNECTION, "upgrade")
+        .header(hyper::header::UPGRADE, "websocket")
+        .header("Sec-WebSocket-Accept", accept_key)
+        .body(Body::empty())
+        .expect("bug: failed to build response");
```

Comment:
I think this is infallible?!

Reply:
It's infallible as long as the headers and such that I use are valid, I think! It's not infallible in the sense of `std::convert::Infallible` or something.